File: src/main/java/dreamdiary/quiz/domain/model/Quiz.java
package dreamdiary.quiz.domain.model;

import dreamdiary.quiz.domain.model.exception.ChoiceNotFoundException;
import dreamdiary.quiz.domain.model.exception.InvalidQuizFormatException;
import dreamdiary.quiz.domain.model.exception.NotSubmitAtException;
import lombok.Builder;
import lombok.Getter;
import org.springframework.util.StringUtils;

import java.time.LocalDateTime;
import java.util.Objects;

@Getter
public class Quiz {
    private final QuizPublicId quizPublicId;
    private final QuizTitle title;
    private final QuizContent content;
    private final Choices choices;
    private final LocalDateTime releaseAt;
    private final LocalDateTime answerReleaseAt;

    @Builder
    public Quiz(final QuizPublicId quizPublicId, final QuizTitle title, final QuizContent content,
                final Choices choices, final LocalDateTime releaseAt, final LocalDateTime answerReleaseAt) {
        if (null == quizPublicId || null == title || null == content) throw new InvalidQuizFormatException();
        if (null == choices || null == releaseAt || null == answerReleaseAt) throw new InvalidQuizFormatException();
        if (!answerReleaseAt.isAfter(releaseAt)) throw new InvalidQuizFormatException();
        this.quizPublicId = quizPublicId;
        this.title = title;
        this.content = content;
        this.choices = choices;
        this.releaseAt = releaseAt;
        this.answerReleaseAt = answerReleaseAt;
    }

    public QuizSubmit submit(final SubmitterUniqId submitterUniqId, final String choicePublicId) {
        if (null == submitterUniqId) throw new InvalidQuizFormatException();
        if (!StringUtils.hasText(choicePublicId)) throw new InvalidQuizFormatException();
        if (this.choices.values().stream().filter(choice -> Objects.equals(choice.publicId(), choicePublicId)).findFirst().isEmpty())
            throw new ChoiceNotFoundException();
        if (this.releaseAt.isAfter(LocalDateTime.now())) throw new NotSubmitAtException();
        return new QuizSubmit(this.quizPublicId, submitterUniqId, choicePublicId);
    }
}

File: applications/quiz/src/main/kotlin/dreamdiary/nineoclock/quiz/domain/model/Quiz.kt
package dreamdiary.nineoclock.quiz.domain.model

import dreamdiary.nineoclock.shard.identifier.ChoicePublicId
import dreamdiary.nineoclock.shard.identifier.QuizId
import java.time.LocalDateTime

internal class Quiz(
    val title: QuizTitle,
    val content: QuizContent,
    val choiceGroup: ChoiceGroup,
    val releaseAt: LocalDateTime,
    val answerReleaseAt: LocalDateTime,
    val id: QuizId = QuizId()
) {
    fun validateAndToSubmit(submitterId: String, choicePublicId: ChoicePublicId): QuizAnswerSubmit {
        val choice = choiceGroup.findChoiceByPublicId(choicePublicId) ?: throw RuntimeException()
        if (releaseAt.isAfter(LocalDateTime.now())) throw RuntimeException()

        return QuizAnswerSubmit(
            submitterId = submitterId,
            quizId = id,
            choiceId = choice.id
        )
    }
}
